allennlp/commands/__init__.py

```diff
@@ -50,59 +50,53 @@ def add_argument(self, *args, **kwargs):
 
 
 def create_parser(
-    prog: str = None, subcommand_overrides: Dict[str, Subcommand] = None
+    prog: str = None, subcommand_overrides: Iterable[Subcommand] = None
 ) -> argparse.ArgumentParser:
     """
     Creates the argument parser for the main program.
     """
-    if subcommand_overrides is None:
-        subcommand_overrides = {}
+    subcommand_overrides = subcommand_overrides or []
 
-    parser = ArgumentParserWithDefaults(description="Run AllenNLP", usage="%(prog)s", prog=prog)
-    parser.add_argument("--version", action="version", version="%(prog)s " + __version__)
+    parser = ArgumentParserWithDefaults(description="Run AllenNLP", prog=prog)
+    parser.add_argument("--version", action="version", version=f"%(prog)s {__version__}")
 
     subparsers = parser.add_subparsers(title="Commands", metavar="")
 
-    subcommands = {
-        # Default commands
-        "train": Train(),
-        "evaluate": Evaluate(),
-        "predict": Predict(),
-        "elmo": Elmo(),
-        "fine-tune": FineTune(),
-        "dry-run": DryRun(),
-        "make-vocab": DryRun(),  # deprecated, but keeping for backward compatibility.
-        "test-install": TestInstall(),
-        "find-lr": FindLearningRate(),
-        "print-results": PrintResults(),
-        # Superseded by overrides
-        **subcommand_overrides,
-    }
-
-    for name, subcommand in subcommands.items():
-        subparser = subcommand.add_subparser(name, subparsers)
-        # configure doesn't need include-package because it imports
-        # whatever classes it needs.
-        if name != "configure":
-            subparser.add_argument(
-                "--include-package",
-                type=str,
-                action="append",
-                default=[],
-                help="additional packages to include",
-            )
+    subcommand_overrides_names = {subcommand.name for subcommand in subcommand_overrides}
+
+    subcommands = set(subcommand_overrides)
+
+    # The class of each instance in the overrides iterable is gonna be in the subclasses
+    # iterable, by definition. So we only add those `Subcommand` subclasses that don't have an
+    # instance in the overrides iterable.
+    for subcommand_class in Subcommand.__subclasses__():
```

Comment:
Checking for `__subclasses__` is not great, and it's brittle (it only looks for one level down, so grandchildren are not included).  Instead, can we just use the `Registry`?  Make all `Subcommands` register themselves with a name, and just look in the registry to get all registered `Subcommands`.  Then you can remove the `overrides` parameter, also, because there's a trivial mechanism to get your command in - just register it.  Also, this would let you remove the `name` property that you added to `Subcommand` - you just use the registered name as the command name, and that's it.

Reply:
> Checking for `__subclasses__` is not great, and it's brittle (it only looks for one level down, so grandchildren are not included).

We could do it recursively.

> Instead, can we just use the `Registry`? Make all `Subcommands` register themselves with a name, and just look in the registry to get all registered `Subcommands`. Then you can remove the overrides parameter, also, because there's a trivial mechanism to get your command in - just register it. Also, this would let you remove the name property that you added to Subcommand - you just use the registered name as the command name, and that's it.

Okay, I'll try with the `Registry`. Makes sense.

Related, I think subclasses (and recursively) could be registered automatically within this kind of `Registrable` API, maybe optionally. See epwalsh/python-registrable#2. I'm not saying to do so here but just saying that, if it's abstracted away then it'd be nice to have that option for the API consumers.